src/System.Linq/src/System/Linq/Intersect.cs

```diff
@@ -19,8 +19,28 @@ public static IEnumerable<TSource> Intersect<TSource>(this IEnumerable<TSource>
             {
                 throw Error.ArgumentNull(nameof(second));
             }
+           
+            IEnumerable<TSource> big = null;
+            IEnumerable<TSource> small = null;
 
-            return IntersectIterator(first, second, null);
+            if (first.Count() > second.Count())
+            {
+                big = first.Distinct();
+                small = second.Distinct();
+            }
+            else
+            {
+                big = second.Distinct();
+                small = first.Distinct();
+            }
+
+            foreach (TSource element in big)
+            {
+                if (small.Contains(element))
```

Comment:
@JonHanna Bear in mind that I came directly to this method because of expecting a not-too-good performance (I saw very slow behaviours in quite a few situations). But I mean you are the boss here, if you don't like the suggestion I cannot do anything. Although I can do more tests tomorrow.